### PR TITLE
Scaffolding tests for `hpandas.compare_dfs()`

### PR DESCRIPTION
**Pull request by tkpratardan**
@gpsaggese @sonaalKant @samarth9008 
This is `stacked PR1` for #682, where unit tests have been added for scaffolding. I have made the changes to the code in unit testing as pointed out [here](https://github.com/sorrentum/sorrentum/pull/717#discussion_r1516865798) and [here](https://github.com/sorrentum/sorrentum/pull/717#discussion_r1516868901)

As I understand, the code would be expected to fail in several test cases with the present version of `hpandas.compare_dfs()` from helpers, as there have been changes to the logic of the code as pointed out [here](https://github.com/sorrentum/sorrentum/pull/717#discussion_r1515296198)  and [here](https://github.com/sorrentum/sorrentum/pull/717#discussion_r1515294959) along with the inclusion of new argument `only_warning`. Also, the assertion messages are expected to be changed when comparing invariants as they are generated by the `pd.testing.assert_frame_equal()` method. 

Please let me know if this is what's expected to be done. 

FYR: 
The workflow for #682 is organized as follows:
1. Main PR: #717 (changes to the main function)
 2. Stacked PR1: #733 (Scaffolding tests for `hpandas.compare_dfs()`)
 4. Stacked PR2: #726 (fixing linter changes in the files being affected, viz., `helpers/hpandas.py` and `helpers/test/test_hpandas.py` )

### CODE REVIEW

File: helpers/test/test_hpandas.py [hunk 1]
```diff
@@ -2465,7 +2468,7 @@ def test7(self) -> None:
 
     def test8(self) -> None:
         """
-        - compare_nans = True
+        compare_nans = True.
```
Comment (sonaalKant) [inline helpers/test/test_hpandas.py:2471]: Remove `.` and add ``` as this is code block. Also you can write a small sentence

File: helpers/test/test_hpandas.py [hunk 2]
```diff
@@ -2481,9 +2484,190 @@ def test8(self) -> None:
...
+        df1 = df2 * adjustment_factor
+        df1.iloc[1, 2] = np.inf
+        with self.assertRaises(AssertionError):
+            hpandas.compare_dfs(
+                df1,
+                df2,
+                diff_mode="pct_change",
+                remove_inf=False,
+                only_warning=False,
+            )
```
Comment (sonaalKant) [inline helpers/test/test_hpandas.py:2635]: Can we do the same as above, assert_equal on assertion raised?

**Review by sonaalKant — CHANGES_REQUESTED** [review]
Good Job!! LGTM. 

Added some nits.

File: helpers/test/test_hpandas.py [hunk 3]
```diff
@@ -2481,9 +2484,190 @@ def test8(self) -> None:
...
+        timestamp
+        2022-01-01 21:01:00+00:00         0.001           0.001            0.001
+        2022-01-01 21:02:00+00:00         0.001           0.001            NaN
+        2022-01-01 21:03:00+00:00         0.001           0.001            0.001
+        """
+        self.assert_equal(actual, expected, fuzzy_match=True)
+
+    def test12(self) -> None:
+        """
+        Check functionality for `remove_inf = True` in presence of `diff_mode =
```
Comment (sonaalKant) [inline helpers/test/test_hpandas.py:2608]: -> `remove_inf = False` right?

**Review by sonaalKant — CHANGES_REQUESTED** [review]

File: helpers/test/test_hpandas.py (hunk 3 above)
Comment (tkpratardan) [reply helpers/test/test_hpandas.py:2608]: Yes, I am sorry for the typo.

**Review by sonaalKant — APPROVED** [review]

**Comment by gpsaggese** [comment]
Some tests are failing due to the new param.
```
helpers/test/test_hpandas.py::Test_compare_dfs::test10 (0.00 s) FAILED            [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test11 (0.00 s) FAILED            [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test12 (0.00 s) FAILED            [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test13 (0.00 s) FAILED            [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test2 (0.01 s) PASSED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test3 (0.01 s) PASSED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test4 (0.01 s) PASSED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test5 (0.01 s) PASSED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test6 (0.01 s) PASSED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test7 (0.01 s) PASSED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test8 (0.00 s) FAILED             [ 43%]
helpers/test/test_hpandas.py::Test_compare_dfs::test9 (0.00 s) PASSED             [ 43%]
```

Pls do a quick fix.

After this, we can merge. Good working unit testing.

File: helpers/test/test_hpandas.py [hunk 4]
```diff
@@ -2484,9 +2484,207 @@ def test8(self) -> None:
...
+        and df2=
+             A  B
+        0  3.0  0
+        1  2.2  0
+        2  NaN  0
+        3  NaN  0
+        4  NaN  0
+        5  inf  0
+        don't have compatible size
+        """
```
Comment (tkpratardan) [inline helpers/test/test_hpandas.py:2519]: @gpsaggese @sonaalKant 
I made changes to the assertion message to be consistent with the changes made [here](https://github.com/sorrentum/sorrentum/pull/717#discussion_r1527426647). Please let me know if this is according to the expectations

**Comment by tkpratardan** [comment]
> Pls do a quick fix.
> 
> After this, we can merge. Good working unit testing.
@gpsaggese 
I am not sure I understand how to do this. Could you please elaborate?

File: helpers/test/test_hpandas.py [hunk 5]
```diff
@@ -2468,7 +2468,7 @@ def test7(self) -> None:
 
     def test8(self) -> None:
         """
-        - compare_nans = True
+        Test NaN comparison with NaNs present at different location in two dataframes with `compare_nans = True`.
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:2471]: This is a good description. You want to highlight what is that you are testing not all the details (that's in the code).

You need to reflow the text to 80 cols (probably the linter does it for you)

File: helpers/test/test_hpandas.py (hunk 4 above)
Comment (gpsaggese) [reply helpers/test/test_hpandas.py:2519]: Yes this looks good

**Review by gpsaggese — COMMENTED** [review]
Good stuff. Let's get all the tests to pass and then we can merge

**Comment by tkpratardan** [comment]
> Good stuff. Let's get all the tests to pass and then we can merge

@gpsaggese 
I am a little confused by your suggestion because I am not sure how to get it done. Certain tests are designed in accordance with the modified code and are expected to fail in the current version on master branch. 
 
However, I have tested it on my local machine, and all the tests did pass, if that's what you mean.

**Comment by gpsaggese** [comment]
> > Good stuff. Let's get all the tests to pass and then we can merge
> 
> @gpsaggese I am a little confused by your suggestion because I am not sure how to get it done. Certain tests are designed in accordance with the modified code and are expected to fail in the current version on master branch.
> 
> However, I have tested it on my local machine, and all the tests did pass, if that's what you mean.

The regressions on GH are reporting a failure https://github.com/sorrentum/sorrentum/actions/runs/8316607993/job/22756287509?pr=733

(We automatically run all the tests after every commit, and we trust only that and not what happens on people's laptops).

FAILED helpers/test/test_hpandas.py::Test_compare_dfs::test10 - TypeError: compare_dfs() got an unexpected keyword argument 'only_warning'
FAILED helpers/test/test_hpandas.py::Test_compare_dfs::test11 - TypeError: compare_dfs() got an unexpected keyword argument 'only_warning'
FAILED helpers/test/test_hpandas.py::Test_compare_dfs::test12 - TypeError: compare_dfs() got an unexpected keyword argument 'only_warning'
FAILED helpers/test/test_hpandas.py::Test_compare_dfs::test13 - TypeError: compare_dfs() got an unexpected keyword argument 'only_warning'
FAILED helpers/test/test_hpandas.py::Test_compare_dfs::test8 - TypeError: compare_dfs() got an unexpected keyword argument 'only_warning'

I would merge master in this branch and run only these tests on your laptop

```
docker> pytest helpers/test/test_hpandas.py::Test_compare_dfs
```

Makes sense? If not you can ping @sonaalKant asking for help

**Comment by tkpratardan** [comment]
@gpsaggese @sonaalKant 
Fixed the issue. please check and verify.

**Review by sonaalKant — APPROVED** [review]
LGTM.

**Review by gpsaggese — COMMENTED** [review]
LG

File: helpers/test/test_hpandas.py [hunk 6]
```diff
@@ -2488,6 +2489,89 @@ def test8(self) -> None:
...
+    def test9(self) -> None:
+        """
+        Test to verify the error when df1 and df2 have different index types.
+        """
+        df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
+        # Create df2 with a DatetimeIndex.
+        dates = pd.date_range("2021-01-01", periods=3)
+        df2 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6], "timestamp": dates})
+        df2 = df2.set_index("timestamp")
+        with self.assertRaises(AssertionError):
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:2501]: We should check the assertion.

We don't want to just look for an assertion without knowing which one

File: helpers/test/test_hpandas.py [hunk 7]
```diff
@@ -2488,6 +2489,89 @@ def test8(self) -> None:
...
+                "tsB": [400, 500, 600],
+                "tsC": [700, 800, 900],
+                "timestamp": timestamp_index,
+            }
+        )
+        df2 = df2.set_index("timestamp")
+        adjustment_factor = 1.00001
+        df1 = df2 * adjustment_factor
+        df1.iloc[1, 2] = np.inf
+        with self.assertRaises(AssertionError):
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:2567]: Same thing

**Review by gpsaggese — COMMENTED** [review]
Let's merge and then follow up.

File: helpers/test/test_hpandas.py (hunk 7 above)
Comment (tkpratardan) [reply helpers/test/test_hpandas.py:2567]: Yes, @sonaalKant and I were planning to redo the changes after the checks have passed.

**Pull request merged**